# Исследование надежности заемщиков


Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.    
Описание данных   
·children — количество детей в семье   
·days_employed — общий трудовой стаж в днях   
·dob_years — возраст клиента в годах   
·education — уровень образования клиента   
·education_id — идентификатор уровня образования   
·family_status — семейное положение   
·family_status_id — идентификатор семейного положения   
·gender — пол клиента   
·income_type — тип занятости   
·debt — имел ли задолженность по возврату кредитов   
·total_income — ежемесячный доход   
·purpose — цель получения кредита   



## Таблица и  общая информацию о данных

**Задание 1. Библиотека pandas. Считаем данные из csv-файла в датафрейм и сохраним в переменную `data`.**

`/datasets/data.csv`

In [ ]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Первые 20 строчек датафрейма `data` .**

In [ ]:
data.head(50)

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42               высшее             0   
1          1    4024.803754         36              среднее             1   
2          0    5623.422610         33              Среднее             1   
3          3    4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0     926.185831         27               высшее             0   
6          0    2879.202052         43               высшее             0   
7          0     152.779569         50              СРЕДНЕЕ             1   
8          2    6929.865299         35               ВЫСШЕЕ             0   
9          0    2188.756445         41              среднее             1   
10         2    4171.483647         36               высшее             0   
11         0     792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0    1846.641941         54  неоконченное высшее             2   
14         0    1844.956182         56               высшее             0   
15         1     972.364419         26              среднее             1   
16         0    1719.934226         35              среднее             1   
17         0    2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0   10038.818549         48              СРЕДНЕЕ             1   
20         1    1311.604166         36              среднее             1   
21         1     253.685166         33              среднее             1   
22         1    1766.644138         24              среднее             1   
23         0     272.981385         21               высшее             0   
24         1  338551.952911         57              среднее             1   
25         0  363548.489348         67              среднее             1   
26         0            NaN         41              среднее             1   
27         0     529.191635         28               высшее             0   
28         1     717.274324         26               высшее             0   
29         0            NaN         63              среднее             1   
30         1  335581.668515         62              среднее             1   
31         0    1682.083438         47            начальное             3   
32         0    4649.910832         34              СРЕДНЕЕ             1   
33         0    1548.637544         48              среднее             1   
34         0    4488.067031         35              среднее             1   
35         0  394021.072184         68              среднее             1   
36         0     176.216688         33              среднее             1   
37         0    6448.810860         43               ВЫСШЕЕ             0   
38         0     597.881827         25               высшее             0   
39         0     650.587796         31               высшее             0   
40         0    1030.362606         30              среднее             1   
41         0            NaN         50              среднее             1   
42         0    1257.496190         20  неоконченное высшее             2   
43         0    4375.681384         43  неоконченное высшее             2   
44         1    1362.041728         26              СРЕДНЕЕ             1   
45         1    1039.451741         49              среднее             1   
46         2    2262.712304         37              среднее             1   
47         1    2689.137274         33               высшее             0   
48         0    3341.067886         45              среднее             1   
49         0    1181.443228         54              среднее             1   

            family_status  family_status_id gender  income_type  debt  \
0  

**Задание 3. Основная информация о датафрейме с помощью метода `info()`.**

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Количество пропущенных значений для каждого столбца. Комбинация двух методов.**

In [ ]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

total_income  catry_days_employed
2.066726e+04  B                      1
2.120528e+04  B                      1
2.136765e+04  E                      1
2.169510e+04  B                      1
2.189561e+04  B                      1
                                    ..
1.711309e+06  D                      1
1.715018e+06  D                      1
1.726276e+06  D                      1
2.200852e+06  E                      1
2.265604e+06  E                      1
Name: catry_days_employed, Length: 19349, dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце мы обработаем на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Заменим все отрицательные значения положительными с помощью метода `abs()`.**

In [ ]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведем медианное значение трудового стажа `days_employed` в днях.**

In [ ]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставим их как есть. Тем более этот столбец не понадобится нам для исследования.

**Задание 8. Выведем перечень уникальных значений столбца `children`.**

In [ ]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалим строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [ ]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведем перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [ ]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполним пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедимся, что все пропуски заполнены. Проверим себя и ещё раз выведим количество пропущенных значений для каждого столбца с помощью двух методов.**

In [ ]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Заменим вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [ ]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработаем неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведем их к нижнему регистру. Проверим остальные столбцы.**

In [ ]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведем на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалим их.**

In [ ]:
data.duplicated().sum()

71

In [ ]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создадим в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000  назначим категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используем собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [ ]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [ ]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведем на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [ ]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

[Текст ссылки](https://)**Задание 18. Создадим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используем собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучим данные в столбце `purpose` и определим, какие подстроки помогут нам правильно определить категорию.**

In [ ]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [ ]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуем данные и отметим зависимости

#### 3.1 Зависимость между количеством детей и возвратом кредита в срок

In [ ]:
data['children'].value_counts()

0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: children, dtype: int64

In [ ]:
data.groupby('children')['debt'].value_counts()# Ваш код будет здесь. Вы можете создавать новые ячейки.

children  debt
0         0       13028
          1        1063
1         0        4364
          1         444
2         0        1858
          1         194
3         0         303
          1          27
4         0          37
          1           4
5         0           9
Name: debt, dtype: int64

In [ ]:
data.groupby('children')['debt'].mean()

children
0    0.075438
1    0.092346
2    0.094542
3    0.081818
4    0.097561
5    0.000000
Name: debt, dtype: float64

**Вывод:**Посмотрел данные подробно разными способами, чтоб получить представление, кого сколько.Выводы на основе последней выборки. Если не учитывать необычное значение 5 детей - 0 процентов, кол-во детей незначительно увеличивает вероятность невозварта кредита в срок.0 детей 0,07/ 1 ребенок - 0,09(2 тыс.) и почти на 0,02 сотых(видимо 2 процента) увеличение, а потом незначительный рост к 4 детям до 0,09(7 тыс.)/чуть снижение до 0,08 при 3 детях. наличие детей незначительно увеличивает вероятность просрочки, правда стоит отметить, что многодетные редко берут кредит

#### 3.2 Зависимость между семейным положением и возвратом кредита в срок

In [ ]:
data['family_status'].value_counts()

женат / замужем          12261
гражданский брак          4134
Не женат / не замужем     2796
в разводе                 1189
вдовец / вдова             951
Name: family_status, dtype: int64

In [ ]:
data.groupby('family_status')['debt'].value_counts()# Ваш код будет здесь. Вы можете создавать новые ячейки.

family_status          debt
Не женат / не замужем  0        2523
                       1         273
в разводе              0        1105
                       1          84
вдовец / вдова         0         888
                       1          63
гражданский брак       0        3749
                       1         385
женат / замужем        0       11334
                       1         927
Name: debt, dtype: int64

In [ ]:
data.groupby('family_status')['debt'].mean()

family_status
Не женат / не замужем    0.097639
в разводе                0.070648
вдовец / вдова           0.066246
гражданский брак         0.093130
женат / замужем          0.075606
Name: debt, dtype: float64

**Вывод:** Гражданский брак и отсутствие семейных обязательств повышают риск просрочки по кредиту, но незначительно на 0,02, на 0,03 по сравнению с минимальным показателем у вдовствующих 0,06.Правда вдовцы не часто берут кредит.

#### 3.3 Зависимость между уровнем дохода и возвратом кредита в срок

In [ ]:
data['total_income_category'].value_counts()# Ваш код будет здесь. Вы можете создавать новые ячейки.

C    15921
B     5014
D      349
A       25
E       22
Name: total_income_category, dtype: int64

In [ ]:
data.groupby('total_income_category')['debt'].value_counts()

total_income_category  debt
A                      0          23
                       1           2
B                      0        4660
                       1         354
C                      0       14568
                       1        1353
D                      0         328
                       1          21
E                      0          20
                       1           2
Name: debt, dtype: int64

In [ ]:
data.groupby('total_income_category')['debt'].mean()

total_income_category
A    0.080000
B    0.070602
C    0.084982
D    0.060172
E    0.090909
Name: debt, dtype: float64

**Вывод:** больше всего вероятность просрочки у людей с минимальным доходом 0,09, но они почти не берут кредит(или им не дают)22.лучше всего отдают кредит 0,06 с доходом ниже среднего, но выше минимального, однако и они нечасто берут кредит 349, а вот люди с доходом выше среднего, но не максимальным чаще берут кредит 5014 и процент невозврата на чуть выше 0,07.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [ ]:
data['purpose_category'].value_counts()

операции с недвижимостью    10751
операции с автомобилем       4279
получение образования        3988
проведение свадьбы           2313
Name: purpose_category, dtype: int64

In [ ]:
data.groupby('purpose_category')['debt'].value_counts()

purpose_category          debt
операции с автомобилем    0       3879
                          1        400
операции с недвижимостью  0       9971
                          1        780
получение образования     0       3619
                          1        369
проведение свадьбы        0       2130
                          1        183
Name: debt, dtype: int64

In [ ]:
data.groupby('purpose_category')['debt'].mean()

purpose_category
операции с автомобилем      0.093480
операции с недвижимостью    0.072551
получение образования       0.092528
проведение свадьбы          0.079118
Name: debt, dtype: float64

**Вывод:** Выше всего риск невозврата операции с автомобилем и получение образования 0,09.Меньше всего на операции с недвижимостью 0,07.

Из 21525 строк 2174 пропуска в стобцах 'days_employed' стаж и 'total_income' доход. Пропущенных значений сравнительно не так много, что в целом позволяет провести общее исследование, т.к. большая часть данных присутствует. Установить точно с чем связаны пропуски нет возможности.  Судя по наблюдениям, скорее всего с человеческим фактором: пенсионер не считает пенсию доходом,кто-то не помнит свой трудовой стаж  и т.п.

### Шаг 4: общий вывод.

Проведено исследование, чтобы выявить влияние различных факторов на вероятность просрочки по выплате кредита. Входные данные предоставляют информацию о клиентах, которая может помочь в установлении связей между различными показателями и платежеспособностью(погашением кредита в срок).

Гипотезы(предположения) исследования о том, что может влиять на погашение кредита в срок:

1).количество детей
2).семейное положение
3).уровень дохода
4).цель кредита.

В результате исследования удалось выявить следующие  зависмости:

1).кредитополучатели без детей являются более ответственными, чем клиенты, у которых есть дети. Так 0 детей 0,07/ 1 ребенок - 0,09(2 тыс.) и почти на 0,02 сотых увеличение, далее незначительный вероятности просрочки продолжается.Видимо, наличие ребенка накладывает на клиента дополнительные финансовое бремя (обязательства) и клиенты либо менее ответственно относятся к своим кредитным обязательствам, либо не имеют возможности в срок оплачивать обязательства.Стоит отметить, что по некоторым группам заемщиков у нас недостаточно данных, так немного информации о тех, у кого 3 и более детей. можно сказать, что выборка не сбалансирована, классы в выборке имеют разные размеры.Необходима большая выборка данных

2).Гражданский брак и отсутствие семейных обязательств повышают риск просрочки по кредиту, на 0,02, на 0,03 по сравнению с минимальным показателем у вдовствующих 0,06.Правда вдовцы не часто берут кредит.Кредитному отделу следует с осторожностью относиться к клиентам, которые не состоят или не состояли в браке. Согласно данным такие клиенты менее ответственные.Люди, которые вступают в узаконенные отношения ведут  более ответственную жизнь, это сказывается и на отношение к своим обязательствам.

3).распределение по доходам: с минимальным доходом 0,09.с доходом ниже среднего, но выше минимального 0,06 , а вот люди с доходом выше среднего, но не максимальным доля невозврата на сотую единицу чуть выше 0,07, а у людей с максимльными доходами 0,08. Вывод:выборка по заемщикам, как с наименьшим доходом, так и с наибольшим крайне недостаточна для формирования однозначных выводов, также и по другим группам сложно установить очевидную зависимость у людей с разными доходами, нужно больше информации

4).влияние цели кредита: Выше всего риск невозврата операции с автомобилем и получение образования 0,09.Меньше всего на операции с недвижимостью 0,07 и проведение свадьбы.  

Итог: выводы из 2 и 4 пунктов можно использовать для принятия решений, для 1 и 3 нужно больше информации, хотя один вывод из 1 гипотезы можно использовать, о том, что факт наличия детей увеличивает вероятность просрочки по кредиту.